File: solutions/queue/ocaml/QueueArray.ml

let init_queue_size = 5

type queue = {
  values : int array;
  q_in   : int;
  q_out  : int
}

let queue_create () =
  {values = Array.make init_queue_size 0; q_in = 0; q_out = 0}

let queue_print queue =
  let rec aux values i =
    if i = queue.q_in then
      Printf.printf "["
    else
      begin
        aux values ((i+1) mod (Array.length queue.values));
        Printf.printf " %d " (values.(i));
        if (i mod (Array.length queue.values)) != queue.q_out then
            Printf.printf ",";
      end
  in
  aux queue.values (queue.q_out);
  Printf.printf "]\n"

let queue_enqueue value queue =
  let new_queue = 
  if ((queue.q_in + 1) mod (Array.length queue.values)) = queue.q_out then
    begin
      let tmp = {values = (Array.make (2*(Array.length queue.values)) 0); q_in = queue.q_in; q_out = if (queue.q_in < queue.q_out) then queue.q_out + (Array.length queue.values) else queue.q_out} in

      if (queue.q_in < queue.q_out) then
        begin
          for i=queue.q_out to Array.length queue.values do
            tmp.values.(i+Array.length queue.values) <- queue.values.(i)
          done;
          for i=0 to queue.q_in do
            tmp.values.(i) <- queue.values.(i)
          done
        end
      else
        for i=queue.q_out to queue.q_in do
          tmp.values.(i) <- queue.values.(i)
        done;
      tmp
    end
  else
    {values = queue.values; q_in = ((queue.q_in +1) mod (Array.length queue.values)); q_out = queue.q_out} in
  new_queue.values.(queue.q_in) <- value;
  new_queue
  
let queue_empty queue =
  queue.q_in = queue.q_out

let queue_peek queue =
  assert (not (queue_empty queue));
  queue.values.(queue.q_out)

(*
let queue_dequeue queue =
  assert (not (queue_empty queue));
  {values=queue.values; q_in = queue.q_in; q_out = (queue.q_out + 1) mod (Array.length queue.values)}
*)

let queue_dequeue queue =
  assert (not (queue_empty queue));
  let new_out = (queue.q_out + 1) mod (Array.length queue.values) in
  let size = Array.length queue.values in
  let num_elements = (queue.q_in - new_out + size) mod size in
  if num_elements < size / 2 && size > init_queue_size then
    let new_size = max (size / 2) init_queue_size in  
    let new_values = Array.make new_size 0 in
    for i = 0 to num_elements - 1 do
      new_values.(i) <- queue.values.((new_out + i) mod size)
    done;
    {values = new_values; q_in = num_elements; q_out = 0}
  else
    {queue with q_out = new_out}

let queue_delete queue =
  queue_create ()

File: solutions/queue/c/queue_array.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <string.h>
#include <stdbool.h>

/* #define MAX_QUEUE_SIZE 100 */
const int MAX_QUEUE_SIZE=100;

/* Structure of the queue */
typedef struct queue {
  int values[MAX_QUEUE_SIZE];
  int in;
  int out;
} queue_s;

/* 
   Creates a new queue and initializes eventual variables
   Parameters : 
   Asserts    : 
   Return     : the pointer to the new empty queue
*/
queue_s *queue_create() {
  queue_s *queue=(queue_s *)malloc(sizeof(queue_s));
  assert(queue != NULL);
  queue->in = 0;
  queue->out = 0;
  return queue;
}

/* 
   Add the given value to the queue
   Parameters : a new value to add
                the address of the current queue
   Asserts    : if the queue already created
                if the queue has enough space
   Return     : the address of the updated queue
*/ 
queue_s *queue_enqueue(int value, queue_s *queue) {
  assert(queue != NULL); /* fail if queue was not created with queue_create */
  assert( (queue->in + 1) % MAX_QUEUE_SIZE != queue->out);
  queue->values[queue->in] = value;
  queue->in = (queue->in+1) % MAX_QUEUE_SIZE;
  return queue;
}

/* 
   Test if the stack queue or not
   Parameters : the address of the current queue
   Asserts    : if the queue is created
   Return     : true if the queue is empty 
                false if the queue is not empty
*/
bool queue_empty(queue_s *queue) {
  assert(queue != NULL); /* fail if queue was not created with queue_create */
  return queue->in == queue->out;
}

/* 
   Read the head element without removing it
   Parameters : the address of the current queue
   Asserts    : if the queue is not empty
   Return     : the value of the head element
*/
int queue_peek(queue_s *queue) {
  assert(!queue_empty(queue));
  return queue->values[queue->out];
}

/* 
   Remove the head element
   Parameters : the address of the current queue
   Asserts    : the queue is not empty
   Return     : the address of the updated queue
*/
queue_s *queue_dequeue(queue_s *queue) {
  assert(!queue_empty(queue));
  queue->out = (queue->out+1)% MAX_QUEUE_SIZE;
  return queue;
}

/* 
   Print the queue element from the head to the last element
   Parameters : the address of the current queue
   Asserts    : if the queue is already created
   Return     : 
*/
void queue_print(queue_s *queue) {
  assert(queue != NULL); /* fail if queue was not created with queue_create */
  printf("[ ");
  for(int i=queue->out; i!=queue->in; i=(i+1)% MAX_QUEUE_SIZE) {
    printf("%d ",queue->values[i]);
    if((i+1)% MAX_QUEUE_SIZE != queue->in)
      printf(", ");
  }
  printf("]");
  return;
}

/* 
   Erase the queue
   Parameters : the address of the current queue
   Asserts    : if the queue is already created
   Return     :
*/
void queue_delete(queue_s *queue) {
  assert(queue != NULL); /* fail if queue was not created with queue_create */
  free(queue);
  return;
}
